File: libcalends/dart/lib/src/calends.dart
import 'dart:ffi';

import 'package:calends/src/calends_exception.dart';

import 'generated_bindings.dart';
import 'util.dart';

class Calends implements Finalizable {
  static void initialize() {
    loadLib().register_panic_handler(Pointer.fromFunction(panicHandler));
  }

  Calends(dynamic stamp, String calendarString, String formatString) {
    final calendar = strToCharPtr(calendarString);
    final format = strToCharPtr(formatString);

    _bound = loadLib();

    if (stamp is String) {
      _reference = _bound.create_string(
        strToCharPtr(stamp),
        calendar,
        format,
      );
    } else if (stamp is int) {
      _reference = _bound.create_long_long(stamp, calendar, format);
    } else if (stamp is double) {
      _reference = _bound.create_double(stamp, calendar, format);
    } else if (stamp is Map<String, String>) {
      if (stamp.containsKey('start') && stamp.containsKey('end')) {
        _reference = _bound.create_string_range(
          strToCharPtr(stamp['start'] as String),
          strToCharPtr(stamp['end'] as String),
          calendar,
          format,
        );
      } else if (stamp.containsKey('start') && stamp.containsKey('duration')) {
        _reference = _bound.create_string_start_period(
          strToCharPtr(stamp['start'] as String),
          strToCharPtr(stamp['duration'] as String),
          calendar,
          format,
        );
      } else if (stamp.containsKey('duration') && stamp.containsKey('end')) {
        _reference = _bound.create_string_end_period(
          strToCharPtr(stamp['duration'] as String),
          strToCharPtr(stamp['end'] as String),
          calendar,
          format,
        );
      }
    } else if (stamp is Map<String, int>) {
      if (stamp.containsKey('start') && stamp.containsKey('end')) {
        _reference = _bound.create_long_long_range(
          stamp['start'] as int,
          stamp['end'] as int,
          calendar,
          format,
        );
      } else if (stamp.containsKey('start') && stamp.containsKey('duration')) {
        _reference = _bound.create_long_long_start_period(
          stamp['start'] as int,
          stamp['duration'] as int,
          calendar,
          format,
        );
      } else if (stamp.containsKey('duration') && stamp.containsKey('end')) {
        _reference = _bound.create_long_long_end_period(
          stamp['duration'] as int,
          stamp['end'] as int,
          calendar,
          format,
        );
      }
    } else if (stamp is Map<String, double>) {
      if (stamp.containsKey('start') && stamp.containsKey('end')) {
        _reference = _bound.create_double_range(
          stamp['start'] as double,
          stamp['end'] as double,
          calendar,
          format,
        );
      } else if (stamp.containsKey('start') && stamp.containsKey('duration')) {
        _reference = _bound.create_double_start_period(
          stamp['start'] as double,
          stamp['duration'] as double,
          calendar,
          format,
        );
      } else if (stamp.containsKey('duration') && stamp.containsKey('end')) {
        _reference = _bound.create_double_end_period(
          stamp['duration'] as double,
          stamp['end'] as double,
          calendar,
          format,
        );
      }
    } else {
      throw CalendsException('Unusable timestamp type');
    }

    _finalizer.attach(this, _reference);
  }

  Calends.decodeText(String textString) {
    _bound = loadLib();
    _reference = _bound.decode_text(strToCharPtr(textString));
    _finalizer.attach(this, _reference);
  }

  Calends.decodeJson(String jsonString) {
    _bound = loadLib();
    _reference = _bound.decode_json(strToCharPtr(jsonString));
    _finalizer.attach(this, _reference);
  }

  Calends._(this._bound, this._reference) {
    _finalizer.attach(this, _reference);
  }

  static final Finalizer<int> _finalizer =
      Finalizer((reference) => loadLib().release(reference));

  late CalendsBindings _bound;

  late int _reference;

  String date(String calendarString, String formatString) {
    final result = _bound.date(
      _reference,
      strToCharPtr(calendarString),
      strToCharPtr(formatString),
    );

    return charPtrToStr(result);
  }

  String duration() {
    final result = _bound.duration(_reference);

    return charPtrToStr(result);
  }

  String endDate(String calendarString, String formatString) {
    final result = _bound.end_date(
      _reference,
      strToCharPtr(calendarString),
      strToCharPtr(formatString),
    );

    return charPtrToStr(result);
  }

  Calends add(dynamic offset, String calendar) {
    Calends retVal;

    switch (offset.runtimeType) {
      case String:
        retVal = Calends._(
          _bound,
          _bound.add_string(
            _reference,
            strToCharPtr(offset as String),
            strToCharPtr(calendar),
          ),
        );
        break;
      case int:
        retVal = Calends._(
          _bound,
          _bound.add_long_long(
            _reference,
            offset as int,
            strToCharPtr(calendar),
          ),
        );
        break;
      case double:
        retVal = Calends._(
          _bound,
          _bound.add_double(
            _reference,
            offset as double,
            strToCharPtr(calendar),
          ),
        );
        break;
      default:
        throw FormatException('offset must be String, int, or double');
    }

    return retVal;
  }

  Calends subtract(dynamic offset, String calendar) {
    Calends retVal;

    switch (offset.runtimeType) {
      case String:
        retVal = Calends._(
          _bound,
          _bound.subtract_string(
            _reference,
            strToCharPtr(offset as String),
            strToCharPtr(calendar),
          ),
        );
        break;
      case int:
        retVal = Calends._(
          _bound,
          _bound.subtract_long_long(
            _reference,
            offset as int,
            strToCharPtr(calendar),
          ),
        );
        break;
      case double:
        retVal = Calends._(
          _bound,
          _bound.subtract_double(
            _reference,
            offset as double,
            strToCharPtr(calendar),
          ),
        );
        break;
      default:
        throw FormatException('offset must be String, int, or double');
    }

    return retVal;
  }

  Calends addFromEnd(dynamic offset, String calendar) {
    Calends retVal;

    switch (offset.runtimeType) {
      case String:
        retVal = Calends._(
          _bound,
          _bound.add_from_end_string(
            _reference,
            strToCharPtr(offset as String),
            strToCharPtr(calendar),
          ),
        );
        break;
      case int:
        retVal = Calends._(
          _bound,
          _bound.add_from_end_long_long(
            _reference,
            offset as int,
            strToCharPtr(calendar),
          ),
        );
        break;
      case double:
        retVal = Calends._(
          _bound,
          _bound.add_from_end_double(
            _reference,
            offset as double,
            strToCharPtr(calendar),
          ),
        );
        break;
      default:
        throw FormatException('offset must be String, int, or double');
    }

    return retVal;
  }

  Calends subtractFromEnd(dynamic offset, String calendar) {
    Calends retVal;

    switch (offset.runtimeType) {
      case String:
        retVal = Calends._(
          _bound,
          _bound.subtract_from_end_string(
            _reference,
            strToCharPtr(offset as String),
            strToCharPtr(calendar),
          ),
        );
        break;
      case int:
        retVal = Calends._(
          _bound,
          _bound.subtract_from_end_long_long(
            _reference,
            offset as int,
            strToCharPtr(calendar),
          ),
        );
        break;
      case double:
        retVal = Calends._(
          _bound,
          _bound.subtract_from_end_double(
            _reference,
            offset as double,
            strToCharPtr(calendar),
          ),
        );
        break;
      default:
        throw FormatException('offset must be String, int, or double');
    }

    return retVal;
  }

  Calends next(dynamic offset, String calendar) {
    Calends retVal;

    switch (offset.runtimeType) {
      case String:
        retVal = Calends._(
          _bound,
          _bound.next_string(
            _reference,
            strToCharPtr(offset as String),
            strToCharPtr(calendar),
          ),
        );
        break;
      case int:
        retVal = Calends._(
          _bound,
          _bound.next_long_long(
            _reference,
            offset as int,
            strToCharPtr(calendar),
          ),
        );
        break;
      case double:
        retVal = Calends._(
          _bound,
          _bound.next_double(
            _reference,
            offset as double,
            strToCharPtr(calendar),
          ),
        );
        break;
      default:
        throw FormatException('offset must be String, int, or double');
    }

    return retVal;
  }

  Calends previous(dynamic offset, String calendar) {
    Calends retVal;

    switch (offset.runtimeType) {
      case String:
        retVal = Calends._(
          _bound,
          _bound.previous_string(
            _reference,
            strToCharPtr(offset as String),
            strToCharPtr(calendar),
          ),
        );
        break;
      case int:
        retVal = Calends._(
          _bound,
          _bound.previous_long_long(
            _reference,
            offset as int,
            strToCharPtr(calendar),
          ),
        );
        break;
      case double:
        retVal = Calends._(
          _bound,
          _bound.previous_double(
            _reference,
            offset as double,
            strToCharPtr(calendar),
          ),
        );
        break;
      default:
        throw FormatException('offset must be String, int, or double');
    }

    return retVal;
  }

  Calends withDate(dynamic stamp, String calendar, String format) {
    int retVal;

    switch (stamp.runtimeType) {
      case String:
        retVal = _bound.with_date_string(
          _reference,
          strToCharPtr(stamp as String),
          strToCharPtr(calendar),
          strToCharPtr(format),
        );
        break;
      case int:
        retVal = _bound.with_date_long_long(
          _reference,
          stamp as int,
          strToCharPtr(calendar),
          strToCharPtr(format),
        );
        break;
      case double:
        retVal = _bound.with_date_double(
          _reference,
          stamp as double,
          strToCharPtr(calendar),
          strToCharPtr(format),
        );
        break;
      default:
        throw FormatException('stamp must be String, int, or double');
    }

    return Calends._(_bound, retVal);
  }

  Calends withEndDate(dynamic stamp, String calendar, String format) {
    int retVal;

    switch (stamp.runtimeType) {
      case String:
        retVal = _bound.with_end_date_string(
          _reference,
          strToCharPtr(stamp as String),
          strToCharPtr(calendar),
          strToCharPtr(format),
        );
        break;
      case int:
        retVal = _bound.with_end_date_long_long(
          _reference,
          stamp as int,
          strToCharPtr(calendar),
          strToCharPtr(format),
        );
        break;
      case double:
        retVal = _bound.with_end_date_double(
          _reference,
          stamp as double,
          strToCharPtr(calendar),
          strToCharPtr(format),
        );
        break;
      default:
        throw FormatException('stamp must be String, int, or double');
    }

    return Calends._(_bound, retVal);
  }

  Calends withDuration(dynamic duration, String calendar) {
    int retVal;

    switch (duration.runtimeType) {
      case String:
        retVal = _bound.with_duration_string(
          _reference,
          strToCharPtr(duration as String),
          strToCharPtr(calendar),
        );
        break;
      case int:
        retVal = _bound.with_duration_long_long(
          _reference,
          duration as int,
          strToCharPtr(calendar),
        );
        break;
      case double:
        retVal = _bound.with_duration_double(
          _reference,
          duration as double,
          strToCharPtr(calendar),
        );
        break;
      default:
        throw FormatException('stamp must be String, int, or double');
    }

    return Calends._(_bound, retVal);
  }

  Calends withDurationFromEnd(dynamic duration, String calendar) {
    int retVal;

    switch (duration.runtimeType) {
      case String:
        retVal = _bound.with_duration_from_end_string(
          _reference,
          strToCharPtr(duration as String),
          strToCharPtr(calendar),
        );
        break;
      case int:
        retVal = _bound.with_duration_from_end_long_long(
          _reference,
          duration as int,
          strToCharPtr(calendar),
        );
        break;
      case double:
        retVal = _bound.with_duration_from_end_double(
          _reference,
          duration as double,
          strToCharPtr(calendar),
        );
        break;
      default:
        throw FormatException('stamp must be String, int, or double');
    }

    return Calends._(_bound, retVal);
  }

  Calends merge(Calends other) {
    return Calends._(
      _bound,
      _bound.merge(_reference, other._reference),
    );
  }

  Calends intersect(Calends other) {
    return Calends._(
      _bound,
      _bound.intersect(_reference, other._reference),
    );
  }

  Calends gap(Calends other) {
    return Calends._(
      _bound,
      _bound.gap(_reference, other._reference),
    );
  }

  String difference(Calends other, String modeString) {
    return charPtrToStr(_bound.difference(
      _reference,
      other._reference,
      strToCharPtr(modeString),
    ));
  }

  int compare(Calends other, String modeString) {
    return _bound.compare(
      _reference,
      other._reference,
      strToCharPtr(modeString),
    );
  }

  bool isSame(Calends other) {
    return _bound.is_same(_reference, other._reference) != 0;
  }

  bool contains(Calends other) {
    return _bound.contains(_reference, other._reference) != 0;
  }

  bool overlaps(Calends other) {
    return _bound.overlaps(_reference, other._reference) != 0;
  }

  bool abuts(Calends other) {
    return _bound.abuts(_reference, other._reference) != 0;
  }

  bool isBefore(Calends other) {
    return _bound.is_before(_reference, other._reference) != 0;
  }

  bool startsBefore(Calends other) {
    return _bound.starts_before(_reference, other._reference) != 0;
  }

  bool endsBefore(Calends other) {
    return _bound.ends_before(_reference, other._reference) != 0;
  }

  bool isDuring(Calends other) {
    return _bound.is_during(_reference, other._reference) != 0;
  }

  bool startsDuring(Calends other) {
    return _bound.starts_during(_reference, other._reference) != 0;
  }

  bool endsDuring(Calends other) {
    return _bound.ends_during(_reference, other._reference) != 0;
  }

  bool isAfter(Calends other) {
    return _bound.is_after(_reference, other._reference) != 0;
  }

  bool startsAfter(Calends other) {
    return _bound.starts_after(_reference, other._reference) != 0;
  }

  bool endsAfter(Calends other) {
    return _bound.ends_after(_reference, other._reference) != 0;
  }

  bool isShorter(Calends other) {
    return _bound.is_shorter(_reference, other._reference) != 0;
  }

  bool isSameDuration(Calends other) {
    return _bound.is_same_duration(_reference, other._reference) != 0;
  }

  bool isLonger(Calends other) {
    return _bound.is_longer(_reference, other._reference) != 0;
  }

  @override
  String toString() {
    final result = _bound.string(_reference);

    return charPtrToStr(result);
  }

  String encodeText() {
    final result = _bound.encode_text(_reference);

    return charPtrToStr(result);
  }

  String encodeJson() {
    final result = _bound.encode_json(_reference);

    return charPtrToStr(result);
  }

  @override
  int get hashCode => _reference.hashCode;

  @override
  bool operator ==(Object other) {
    return other is Calends && isSame(other);
  }

  bool operator <(Calends other) {
    return isBefore(other);
  }

  bool operator <=(Calends other) {
    return isBefore(other) || isSame(other);
  }

  bool operator >=(Calends other) {
    return isAfter(other) || isSame(other);
  }

  bool operator >(Calends other) {
    return isAfter(other);
  }
}

File: libcalends/php/src/Calends.php
<?php

namespace Calends;

require_once('CalendsException.php');

class Calends implements \JsonSerializable
{
    private static ?\FFI $ffi = null;

    private ?int $instance = null;

    private static function ffiInit(): void
    {
        if (is_null(self::$ffi)) {
            self::$ffi = \FFI::scope("CALENDS");
        }
    }

    private function __construct(int $instance)
    {
        if (empty($instance)) {
            throw CalendsException::current();
        }

        $this->instance = $instance;
    }

    public function __destruct()
    {
        if (!empty($this->instance)) {
            self::$ffi->Calends_release($this->instance);
        } elseif (!empty(CalendsException::current())) {
            throw CalendsException::current();
        }
    }

    // CREATE

    public static function create($stamp = '', string $calendar = '', string $format = ''): self
    {
        self::ffiInit();

        if (is_string($stamp)) {
            return new self((int)self::$ffi->Calends_create_string($stamp, $calendar, $format));
        } elseif (is_int($stamp)) {
            return new self((int)self::$ffi->Calends_create_long_long($stamp, $calendar, $format));
        } elseif (is_float($stamp)) {
            return new self((int)self::$ffi->Calends_create_double($stamp, $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('start', $stamp) && array_key_exists('end', $stamp) && is_string($stamp['start'])) {
            return new self((int)self::$ffi->Calends_create_string_range($stamp['start'], $stamp['end'], $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('start', $stamp) && array_key_exists('duration', $stamp) && is_string($stamp['start'])) {
            return new self((int)self::$ffi->Calends_create_string_start_period($stamp['start'], $stamp['duration'], $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('duration', $stamp) && array_key_exists('end', $stamp) && is_string($stamp['end'])) {
            return new self((int)self::$ffi->Calends_create_string_end_period($stamp['duration'], $stamp['end'], $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('start', $stamp) && array_key_exists('end', $stamp) && is_int($stamp['start'])) {
            return new self((int)self::$ffi->Calends_create_long_long_range($stamp['start'], $stamp['end'], $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('start', $stamp) && array_key_exists('duration', $stamp) && is_int($stamp['start'])) {
            return new self((int)self::$ffi->Calends_create_long_long_start_period($stamp['start'], $stamp['duration'], $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('duration', $stamp) && array_key_exists('end', $stamp) && is_int($stamp['end'])) {
            return new self((int)self::$ffi->Calends_create_long_long_end_period($stamp['duration'], $stamp['end'], $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('start', $stamp) && array_key_exists('end', $stamp) && is_float($stamp['start'])) {
            return new self((int)self::$ffi->Calends_create_double_range($stamp['start'], $stamp['end'], $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('start', $stamp) && array_key_exists('duration', $stamp) && is_float($stamp['start'])) {
            return new self((int)self::$ffi->Calends_create_double_start_period($stamp['start'], $stamp['duration'], $calendar, $format));
        } elseif (is_array($stamp) && array_key_exists('duration', $stamp) && array_key_exists('end', $stamp) && is_float($stamp['end'])) {
            return new self((int)self::$ffi->Calends_create_double_end_period($stamp['duration'], $stamp['end'], $calendar, $format));
        } else {
            throw new CalendsException('Unsupported type');
        }
    }

    // READ

    public function date(string $calendar = '', string $format = ''): string
    {
        return \FFI::string(self::$ffi->Calends_date($this->instance, $calendar, $format));
    }

    public function duration(): string
    {
        return \FFI::string(self::$ffi->Calends_duration($this->instance));
    }

    public function endDate(string $calendar = '', string $format = ''): string
    {
        return \FFI::string(self::$ffi->Calends_end_date($this->instance, $calendar, $format));
    }

    // UPDATE

    public function add($offset = '', string $calendar = ''): self
    {
        if (is_string($offset)) {
            return new self(self::$ffi->Calends_add_string($this->instance, $offset, $calendar));
        } elseif (is_int($offset)) {
            return new self(self::$ffi->Calends_add_long_long($this->instance, $offset, $calendar));
        } elseif (is_float($offset)) {
            return new self(self::$ffi->Calends_add_double($this->instance, $offset, $calendar));
        }
    }

    public function subtract($offset = '', string $calendar = ''): self
    {
        if (is_string($offset)) {
            return new self(self::$ffi->Calends_subtract_string($this->instance, $offset, $calendar));
        } elseif (is_int($offset)) {
            return new self(self::$ffi->Calends_subtract_long_long($this->instance, $offset, $calendar));
        } elseif (is_float($offset)) {
            return new self(self::$ffi->Calends_subtract_double($this->instance, $offset, $calendar));
        }
    }

    public function addFromEnd($offset = '', string $calendar = ''): self
    {
        if (is_string($offset)) {
            return new self(self::$ffi->Calends_add_from_end_string($this->instance, $offset, $calendar));
        } elseif (is_int($offset)) {
            return new self(self::$ffi->Calends_add_from_end_long_long($this->instance, $offset, $calendar));
        } elseif (is_float($offset)) {
            return new self(self::$ffi->Calends_add_from_end_double($this->instance, $offset, $calendar));
        }
    }

    public function subtractFromEnd($offset = '', string $calendar = ''): self
    {
        if (is_string($offset)) {
            return new self(self::$ffi->Calends_subtract_from_end_string($this->instance, $offset, $calendar));
        } elseif (is_int($offset)) {
            return new self(self::$ffi->Calends_subtract_from_end_long_long($this->instance, $offset, $calendar));
        } elseif (is_float($offset)) {
            return new self(self::$ffi->Calends_subtract_from_end_double($this->instance, $offset, $calendar));
        }
    }

    public function next($offset = '', string $calendar = ''): self
    {
        if (is_string($offset)) {
            return new self(self::$ffi->Calends_next_string($this->instance, $offset, $calendar));
        } elseif (is_int($offset)) {
            return new self(self::$ffi->Calends_next_long_long($this->instance, $offset, $calendar));
        } elseif (is_float($offset)) {
            return new self(self::$ffi->Calends_next_double($this->instance, $offset, $calendar));
        }
    }

    public function previous($offset = '', string $calendar = ''): self
    {
        if (is_string($offset)) {
            return new self(self::$ffi->Calends_previous_string($this->instance, $offset, $calendar));
        } elseif (is_int($offset)) {
            return new self(self::$ffi->Calends_previous_long_long($this->instance, $offset, $calendar));
        } elseif (is_float($offset)) {
            return new self(self::$ffi->Calends_previous_double($this->instance, $offset, $calendar));
        }
    }

    public function withDate($stamp = '', string $calendar = '', string $format = ''): self
    {
        if (is_string($stamp)) {
            return new self(self::$ffi->Calends_with_date_string($this->instance, $stamp, $calendar, $format));
        } elseif (is_int($stamp)) {
            return new self(self::$ffi->Calends_with_date_long_long($this->instance, $stamp, $calendar, $format));
        } elseif (is_float($stamp)) {
            return new self(self::$ffi->Calends_with_date_double($this->instance, $stamp, $calendar, $format));
        }
    }

    public function withEndDate($stamp = '', string $calendar = '', string $format = ''): self
    {
        if (is_string($stamp)) {
            return new self(self::$ffi->Calends_with_end_date_string($this->instance, $stamp, $calendar, $format));
        } elseif (is_int($stamp)) {
            return new self(self::$ffi->Calends_with_end_date_long_long($this->instance, $stamp, $calendar, $format));
        } elseif (is_float($stamp)) {
            return new self(self::$ffi->Calends_with_end_date_double($this->instance, $stamp, $calendar, $format));
        }
    }

    public function withDuration($offset = '', string $calendar = ''): self
    {
        if (is_string($offset)) {
            return new self(self::$ffi->Calends_with_duration_string($this->instance, $offset, $calendar));
        } elseif (is_int($offset)) {
            return new self(self::$ffi->Calends_with_duration_long_long($this->instance, $offset, $calendar));
        } elseif (is_float($offset)) {
            return new self(self::$ffi->Calends_with_duration_double($this->instance, $offset, $calendar));
        }
    }

    public function withDurationFromEnd($offset = '', string $calendar = ''): self
    {
        if (is_string($offset)) {
            return new self(self::$ffi->Calends_with_duration_from_end_string($this->instance, $offset, $calendar));
        } elseif (is_int($offset)) {
            return new self(self::$ffi->Calends_with_duration_from_end_long_long($this->instance, $offset, $calendar));
        } elseif (is_float($offset)) {
            return new self(self::$ffi->Calends_with_duration_from_end_double($this->instance, $offset, $calendar));
        }
    }

    public function merge(self $other): self
    {
        return new self(self::$ffi->Calends_merge($this->instance, $other->instance));
    }

    public function intersect(self $other): self
    {
        return new self(self::$ffi->Calends_intersect($this->instance, $other->instance));
    }

    public function gap(self $other): self
    {
        return new self(self::$ffi->Calends_gap($this->instance, $other->instance));
    }

    // COMPARISON

    public function difference(self $other, string $mode = ''): string
    {
        return \FFI::string(self::$ffi->Calends_difference($this->instance, $other->instance, $mode));
    }

    public function compare(self $other, string $mode = ''): int
    {
        return (int)self::$ffi->Calends_compare($this->instance, $other->instance, $mode);
    }

    public function isSame(self $other): bool
    {
        return (int)self::$ffi->Calends_is_same($this->instance, $other->instance);
    }

    public function isSameDuration(self $other): bool
    {
        return (int)self::$ffi->Calends_is_same_duration($this->instance, $other->instance);
    }

    public function isShorter(self $other): bool
    {
        return (int)self::$ffi->Calends_is_shorter($this->instance, $other->instance);
    }

    public function isLonger(self $other): bool
    {
        return (int)self::$ffi->Calends_is_longer($this->instance, $other->instance);
    }

    public function contains(self $other): bool
    {
        return (int)self::$ffi->Calends_contains($this->instance, $other->instance);
    }

    public function overlaps(self $other): bool
    {
        return (int)self::$ffi->Calends_overlaps($this->instance, $other->instance);
    }

    public function abuts(self $other): bool
    {
        return (int)self::$ffi->Calends_abuts($this->instance, $other->instance);
    }

    public function isBefore(self $other): bool
    {
        return (int)self::$ffi->Calends_is_before($this->instance, $other->instance);
    }

    public function startsBefore(self $other): bool
    {
        return (int)self::$ffi->Calends_starts_before($this->instance, $other->instance);
    }

    public function endsBefore(self $other): bool
    {
        return (int)self::$ffi->Calends_ends_before($this->instance, $other->instance);
    }

    public function isDuring(self $other): bool
    {
        return (int)self::$ffi->Calends_is_during($this->instance, $other->instance);
    }

    public function startsDuring(self $other): bool
    {
        return (int)self::$ffi->Calends_starts_during($this->instance, $other->instance);
    }

    public function endsDuring(self $other): bool
    {
        return (int)self::$ffi->Calends_ends_during($this->instance, $other->instance);
    }

    public function isAfter(self $other): bool
    {
        return (int)self::$ffi->Calends_is_after($this->instance, $other->instance);
    }

    public function startsAfter(self $other): bool
    {
        return (int)self::$ffi->Calends_starts_after($this->instance, $other->instance);
    }

    public function endsAfter(self $other): bool
    {
        return (int)self::$ffi->Calends_ends_after($this->instance, $other->instance);
    }

    // UTILITY

    public function __toString(): string
    {
        return \FFI::string(self::$ffi->Calends_string($this->instance));
    }

    public static function jsonUnserialize(string $json): self
    {
        self::ffiInit();

        return new self((int)self::$ffi->Calends_decode_json($json));
    }

    public function jsonSerialize(): array
    {
        try {
            $out = json_decode(\FFI::string(self::$ffi->Calends_encode_json($this->instance)), true);
        } catch (\TypeError $typeError) {
            $exception = CalendsException::current();

            if (empty($exception)) {
                throw $typeError;
            }

            throw $exception;
        }

        return is_array($out) ? $out : ['start' => $out, 'end' => $out];
    }

    public function __unserialize(array $text): void
    {
        self::ffiInit();

        $this->instance = ((int)self::$ffi->Calends_decode_text($text[0]));
    }

    public function __serialize(): array
    {
        return [\FFI::string(self::$ffi->Calends_encode_text($this->instance))];
    }

    public function __debugInfo(): array
    {
        return $this->jsonSerialize();
    }
}
